### ISSUE TITLE
e2e flake: Restart [Disruptive] should restart all nodes and ensure all nodes and pods recover

### ISSUE THREAD

**Issue by davidopp** [kubernetes/kubernetes#22489]
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-serial/741/
http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-serial/733/

**Comment by davidopp** [comment]
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/restart.go:116
Expected error:
    <*errors.errorString | 0xc20af190d0>: {
        s: "couldn't find 13 pods within 5m0s; last error: expected to find 13 pods but found only 14",
    }
    couldn't find 13 pods within 5m0s; last error: expected to find 13 pods but found only 14
not to have occurred
```

cc/ @dchen1107 @gmarek

**Comment by lavalamp** [comment]
@krousey can you dig through the logs and say what component didn't come up? and/or was it because a node failed to come up?

**Comment by krousey** [comment]
@lavalamp I can look into it today, but I'm on vacation for a week after this. I'll update with what I can and then un-assign myself at the end of the day.

**Comment by krousey** [comment]
@davidopp Please remember to include links to results in storage. These are public and last much longer than the Jenkins results.

https://console.developers.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce-serial/741/
https://console.developers.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce-serial/733/

**Comment by krousey** [comment]
So after a restart we see more pods than we had before. So is the idea that something failed the first time and worked after a restart? Still combing through to correlate what pods are where.

**Comment by krousey** [comment]
Looking at the 733 build, this is what was running prior to restart:

```
Mar  2 20:00:39.058: INFO: Waiting up to 2m0s for pod elasticsearch-logging-v1-6p93j status to be running and ready
Mar  2 20:00:39.058: INFO: Waiting up to 2m0s for pod monitoring-influxdb-grafana-v3-q0ggo status to be running and ready
Mar  2 20:00:39.058: INFO: Waiting up to 2m0s for pod fluentd-elasticsearch-jenkins-e2e-minion-7dbv status to be running and ready
Mar  2 20:00:39.058: INFO: Waiting up to 2m0s for pod heapster-v14-kuekp status to be running and ready
Mar  2 20:00:39.059: INFO: Waiting up to 2m0s for pod kube-proxy-jenkins-e2e-minion-7dbv status to be running and ready
Mar  2 20:00:39.059: INFO: Waiting up to 2m0s for pod elasticsearch-logging-v1-35g6f status to be running and ready
Mar  2 20:00:39.059: INFO: Waiting up to 2m0s for pod kube-dns-v10-e94ih status to be running and ready
Mar  2 20:00:39.059: INFO: Waiting up to 2m0s for pod kubernetes-dashboard-v1.0.0beta1-libok status to be running and ready
Mar  2 20:00:39.059: INFO: Waiting up to 2m0s for pod kube-proxy-jenkins-e2e-minion-v77p status to be running and ready
Mar  2 20:00:39.059: INFO: Waiting up to 2m0s for pod l7-lb-controller-v0.5.2-os4k0 status to be running and ready
Mar  2 20:00:39.059: INFO: Waiting up to 2m0s for pod kibana-logging-v1-k75xt status to be running and ready
Mar  2 20:00:39.059: INFO: Waiting up to 2m0s for pod fluentd-elasticsearch-jenkins-e2e-minion-v77p status to be running and ready
```

And this is what was running after:

```
STEP: Collecting events from namespace "e2e-tests-restart-yen28".                                         
Mar  2 20:06:25.071: INFO: POD                                            NODE                     PHASE  
Mar  2 20:06:25.071: INFO: elasticsearch-logging-v1-35g6f                 jenkins-e2e-minion-v77p  Running
Mar  2 20:06:25.081: INFO: elasticsearch-logging-v1-6p93j                 jenkins-e2e-minion-7dbv  Running
Mar  2 20:06:25.081: INFO: fluentd-elasticsearch-jenkins-e2e-minion-7dbv  jenkins-e2e-minion-7dbv  Running
Mar  2 20:06:25.081: INFO: fluentd-elasticsearch-jenkins-e2e-minion-b4zy  jenkins-e2e-minion-b4zy  Running
Mar  2 20:06:25.081: INFO: fluentd-elasticsearch-jenkins-e2e-minion-v77p  jenkins-e2e-minion-v77p  Running
Mar  2 20:06:25.081: INFO: heapster-v14-kuekp                             jenkins-e2e-minion-v77p  Running
Mar  2 20:06:25.081: INFO: kibana-logging-v1-k75xt                        jenkins-e2e-minion-7dbv  Running
Mar  2 20:06:25.081: INFO: kube-dns-v10-e94ih                             jenkins-e2e-minion-v77p  Running
Mar  2 20:06:25.081: INFO: kube-proxy-jenkins-e2e-minion-7dbv             jenkins-e2e-minion-7dbv  Running
Mar  2 20:06:25.081: INFO: kube-proxy-jenkins-e2e-minion-b4zy             jenkins-e2e-minion-b4zy  Running
Mar  2 20:06:25.081: INFO: kube-proxy-jenkins-e2e-minion-v77p             jenkins-e2e-minion-v77p  Running
Mar  2 20:06:25.081: INFO: kubernetes-dashboard-v1.0.0beta1-libok         jenkins-e2e-minion-v77p  Running
Mar  2 20:06:25.081: INFO: l7-lb-controller-v0.5.2-os4k0                  jenkins-e2e-minion-7dbv  Running
Mar  2 20:06:25.081: INFO: monitoring-influxdb-grafana-v3-q0ggo           jenkins-e2e-minion-7dbv  Running
```

So...

Pod | Node (after) | Existed before
--- | --- | ---
fluentd-elasticsearch-jenkins-e2e-minion-b4zy | jenkins-e2e-minion-b4zy | :bangbang:
kube-proxy-jenkins-e2e-minion-b4zy   |          jenkins-e2e-minion-b4zy | :bangbang: 
fluentd-elasticsearch-jenkins-e2e-minion-v77p | jenkins-e2e-minion-v77p | :white_check_mark: 
elasticsearch-logging-v1-35g6f           |      jenkins-e2e-minion-v77p |  :white_check_mark: 
heapster-v14-kuekp                        |     jenkins-e2e-minion-v77p | :white_check_mark: 
kube-dns-v10-e94ih                         |    jenkins-e2e-minion-v77p | :white_check_mark: 
kube-proxy-jenkins-e2e-minion-v77p    |         jenkins-e2e-minion-v77p | :white_check_mark: 
kubernetes-dashboard-v1.0.0beta1-libok |        jenkins-e2e-minion-v77p | :white_check_mark: 
elasticsearch-logging-v1-6p93j            |     jenkins-e2e-minion-7dbv | :white_check_mark: 
fluentd-elasticsearch-jenkins-e2e-minion-7dbv |  jenkins-e2e-minion-7dbv | :white_check_mark: 
kibana-logging-v1-k75xt                  |      jenkins-e2e-minion-7dbv | :white_check_mark: 
kube-proxy-jenkins-e2e-minion-7dbv  |           jenkins-e2e-minion-7dbv | :white_check_mark: 
l7-lb-controller-v0.5.2-os4k0               |   jenkins-e2e-minion-7dbv | :white_check_mark: 
monitoring-influxdb-grafana-v3-q0ggo |          jenkins-e2e-minion-7dbv | :white_check_mark: 

So I'm going to look more into what happened on jekins-e2e-minion-b4zy.

**Comment by krousey** [comment]
So as best as I can guess
 1. jenkins-e2e-minion-b4zy reported a ready status
 2. the apiserver listed 12 pods
 3. the test waited for those 12 pods to become ready
 4. the test rebooted the nodes
 5. jenkins-e2e-minion-b4zy came back with 2 pods that weren't listed before

Where do these pods come from? A manifest file? Daemon set? Is there a race between a node being reported as ready, the system noticing it, and the system pods being scheduled on it? If so, this test needs to take that into account.

**Comment by davidopp** [comment]
@gmarek do you think you could take a look at this? (@krousey is going on vacation)

**Comment by gmarek** [comment]
Not this week - I have a presentation to prepare.

**Comment by mml** [comment]
I'll try to take a look tomorrow.

**Comment by bprashanth** [comment]
Pods not exisintg before and existing after seems to be what I also obseved in http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-serial/741/consoleFull

This could very likely be because the node on which the kubeproxy didn't start (jenkins-e2e-minion-x6vq) was added to the cluster  ~10m ago in the "should be able to delete nodes" test.

The "should be able to delete nodes" test found 14 pods, but there's a chance it was looking up pods on some old node (i.e it deleted the node and the node controller didn't delete the pods yet).

**Comment by bprashanth** [comment]
>Where do these pods come from? A manifest file? Daemon set? 

fluentd and kube-proxy are static pods, the kubelet reads manifests, starts docker containers and creates associated pod objects in the apiserver to reflect these manifest-static-pods. We're polling and finding/not finding the reflected mirror pods.

**Comment by mml** [comment]
The test has a broken assumption: that the pods it lists once it gets to the "ensuring all pods are running and ready" step are all the interesting pods.  It will wait and poll to see if *that list* of pods gets healthy, and of course sometimes that list won't be complete.

Later, the test passes if the new set of pods acquired by the same method (a) has the same length, and (b) all get healthy within the timeout.

I think we should have a built-in assumption about how many pods we expect to see, and maybe even patterns that match the correct names.  The test will break when we add or remove a mirror pod or daemon set, but it will be much more robust and correct. @kubernetes/sig-testing any objections to this pattern?

**Comment by mml** [comment]
Hmm.. unfortunately checkPodsRunningReady() takes a list of strings.  A simpler approach is to loop over listing all the pods until that list has the length we expect (14 AFAICT).  Until we have a list of 14 pod names, we won't enter the phase where we're waiting for those pods to be running and ready.

**Comment by mml** [comment]
#22831 is a fix in progress.

**Comment by mml** [comment]
@davidopp should this be milestoned to 1.2?

**Comment by davidopp** [comment]
I guess it's moot now since the PR wasn't reviewed in time.

**Comment by davidopp** [comment]
(Sorry for not seeing this sooner.)

**Comment by mml** [comment]
We aren't cherrypicking bug fixes into 1.2 any more?

**Comment by lavalamp** [comment]
Closing due to age. New occurrences will be tracked in auto-filed issues.
